checker/bin/wpi.sh

```diff
@@ -0,0 +1,175 @@
+#!/bin/bash
+
+# This script performs whole-program inference on a project directory.
+
+# For usage and requirements, see the "Whole-program inference"
+# section of the Checker Framework manual:
+# https://checkerframework.org/manual/#whole-program-inference
+
+
+while getopts "d:t:" opt; do
+  case $opt in
+    d) DIR="$OPTARG"
+       ;;
+    t) TIMEOUT="$OPTARG"
+       ;;
+    \?) # echo "Invalid option -$OPTARG" >&2
+       ;;
+  esac
+done
+
+# Make $@ be the arguments that should be passed to dljc.
+shift $(( OPTIND - 1 ))
+
+# check required arguments and environment variables:
+
+# Testing for JAVA8_HOME, not a misspelling of JAVA_HOME.
+# shellcheck disable=SC2153
+if [ "x${JAVA8_HOME}" = "x" ]; then
+    echo "JAVA8_HOME must be set to a Java 8 JDK"
+    exit 1
+fi
+
+if [ ! -d "${JAVA8_HOME}" ]; then
+    echo "JAVA8_HOME is set to a non-existent directory ${JAVA8_HOME}"
+    exit 1
+fi
+
+# Testing for JAVA11_HOME, not a misspelling of JAVA_HOME.
+# shellcheck disable=SC2153
+if [ "x${JAVA11_HOME}" = "x" ]; then
+    echo "JAVA11_HOME must be set to a Java 11 JDK"
+    exit 1
+fi
+
+if [ ! -d "${JAVA11_HOME}" ]; then
+    echo "JAVA11_HOME is set to a non-existent directory ${JAVA11_HOME}"
+    exit 1
+fi
+
+JAVA_HOME="${JAVA11_HOME}"
+
+if [ "x${CHECKERFRAMEWORK}" = "x" ]; then
+    echo "CHECKERFRAMEWORK is not set; it must be set to a locally-built Checker Framework. Please clone and build github.com/typetools/checker-framework"
+    exit 2
+fi
+
+if [ ! -d "${CHECKERFRAMEWORK}" ]; then
+    echo "CHECKERFRAMEWORK is set to a non-existent directory ${CHECKERFRAMEWORK}"
+    exit 2
+fi
+
+if [ "x${DIR}" = "x" ]; then
+    echo "wpi.sh was called without a -d argument. Running on the current directory instead."
+    DIR=$(pwd)
+fi
+
+if [ ! -d "${DIR}" ]; then
+    echo "wpi.sh's -d argument was not a directory: ${DIR}"
+    exit 4
+fi
+
+function configure_and_exec_dljc {
+
+  if [ -f build.gradle ]; then
+      if [ -f gradlew ]; then
+    chmod +x gradlew
+    GRADLE_EXEC="./gradlew"
+      else
+    GRADLE_EXEC="gradle"
+      fi
+      CLEAN_CMD="${GRADLE_EXEC} clean -g .gradle -Dorg.gradle.java.home=${JAVA_HOME}"
```

Comment:
The `-g .gradle` command causes gradle to core dump, when the .gradle directory does not exist.  Is it needed?  Or, can it be supplied only if a `.gradle/` directory exists?

Reply:
As we discussed on our call, I intended to just guard this behind a flag. But I did some digging and now I remember why this is here. There is a relevant Gradle issue (https://github.com/gradle/gradle/issues/1319), which I'll link from a TODO comment in the script. The gist is that Gradle doesn't play nicely with distributed file systems (like those used on the cycle servers) because it tries to put its caches in the user's home directory by default (which can cause either contention if you're using multiple servers or use too much space. And in the cycle servers case, it's also slow! The `-g` option controls where those caches go, in addition to controlling where to look first for user configuration etc. So this is here to force gradle to cache everything in the `.gradle` directory relative to the project, which means each project gets its own cache.

I've implemented a check for whether `.gradle` exists, and create it if not. This leaves the gradle command line invocation unchanged. This should fix your immediate problem, but it leaves the larger problem of not sharing caches; I'm not sure if we want to fix that or not.

Reply:
If it is convenient to do, I would prefer to use a location in /scratch/$USER/ or /tmp/$USER/.  This would permit sharing the cache and would reduce repeated downloads.  This isn't critical but would be nice to do.

Reply:
That would work on the CSE cycle servers, but these scripts should be portable and we can't know a priori what's part of a shared file system and what is not. For that reason, I think it's better to default this to a directory that's exclusive to the project, even if it makes these scripts less efficient.

You're right that on a system like the CSE file servers it would be nice to have an option that will change where this default is. I've added one: `-g`. Hopefully that will address your concerns.